osu.Game/Overlays/BeatmapSet/BeatmapBadgePill.cs

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Extensions.LocalisationExtensions;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Localisation;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+
+#nullable enable
+
+namespace osu.Game.Overlays.BeatmapSet
+{
+    public abstract class BeatmapBadgePill : CompositeDrawable
+    {
+        [Resolved]
+        protected OsuColour Colours { get; private set; } = null!;
+
+        [Resolved(canBeNull: true)]
+        protected OverlayColourProvider? ColourProvider { get; private set; }
+
+        /// <summary>
+        /// The text displayed on the badge's label.
+        /// </summary>
+        public abstract LocalisableString BadgeText { get; }
+
+        /// <summary>
+        /// The colour of the badge's label.
+        /// </summary>
+        public abstract Colour4 BadgeColour { get; }
```

Comment:
Can these be plain settable properties rather than abstract get-only? If I'm honest I really cooled off on those get-only abstract/virtual ones after the number of times they cause trouble.

Reply:
Don't see the big problem with them? it's very unlikely for this `BeatmapBadgePill` to be inherited multiple levels deeper for virtual properties to become a potential issue. Unless you mean in regards to DI?

Reply:
they create complications, especially when mixed with DI'd properties. now the override working depends on when the component is calling the property, it depends on DI ordering. while the expectation from a plain settable property is that it should work always, ever, period.

Reply:
Fair, will change.